### PR TITLE
Make TestCluster#wipe run steps in parallel where possible

### PR DESCRIPTION
**Pull request by original-brownbear**
This is one of the most expensive methods in our tests.
We don't have to execute the steps sequentially here, so we can speed things up considerably by only making things that need to wait for each other wait.
Also I believe this does add a certain degree of extra coverage on the network layer as we have very few tests that run parallel requests, so we get to stress the allocation logic and such a little extra here.

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java [hunk 1]
```diff
@@ -136,52 +235,93 @@ public void assertAfterTest() throws Exception {
...
+        client().admin().indices().prepareGetTemplates().execute(listener.delegateFailure((f, response) -> {
+            try (RefCountingListener subscribableListener = new RefCountingListener(f)) {
+                for (IndexTemplateMetadata indexTemplate : response.getIndexTemplates()) {
+                    if (exclude.contains(indexTemplate.getName())) {
+                        continue;
+                    }
+                    client().admin()
+                        .indices()
+                        .prepareDeleteTemplate(indexTemplate.getName())
+                        .execute(subscribableListener.acquire().<AcknowledgedResponse>safeMap(r -> {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:311]: Maybe use the `acquire(CheckedConsumer<>)` override?

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java [hunk 2]
```diff
@@ -136,52 +235,93 @@ public void assertAfterTest() throws Exception {
...
-            for (IndexTemplateMetadata indexTemplate : response.getIndexTemplates()) {
-                if (exclude.contains(indexTemplate.getName())) {
-                    continue;
-                }
-                try {
-                    client().admin().indices().prepareDeleteTemplate(indexTemplate.getName()).get();
-                } catch (IndexTemplateMissingException e) {
-                    // ignore
+    private void wipeAllTemplates(Set<String> exclude, ActionListener<Void> listener) {
+        client().admin().indices().prepareGetTemplates().execute(listener.delegateFailure((f, response) -> {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:302]: Suggest using `SubscribableListener#andThen` here. I mean I know it's only a chain of 2 things, but still, saves some nesting.

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java [hunk 3]
```diff
@@ -136,52 +235,93 @@ public void assertAfterTest() throws Exception {
      * all indices are removed.
      */
     public void wipeIndices(String... indices) {
+        final PlainActionFuture<Void> future = new PlainActionFuture<>();
+        wipeIndicesAsync(indices, future);
+        try {
+            future.get();
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:241]: Could we use a `safeGet` or `safeAwait` variant? Or is 10s not enough here? In any case I'd rather a bounded wait and we should restore the interrupted flag if set.

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java [hunk 4]
```diff
@@ -69,14 +71,111 @@ public void beforeTest(Random randomGenerator) throws IOException, InterruptedEx
...
+                            delegate.onFailure(e);
+                        } else {
+                            onResponse(AcknowledgedResponse.TRUE);
+                        }
+                    }
+                }
+            );
+        }
+        try {
+            done.get();
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:112]: Likewise here, can we bound the wait and handle interrupts properly?

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java [hunk 5]
```diff
@@ -136,52 +235,93 @@ public void assertAfterTest() throws Exception {
...
-                if (exclude.contains(indexTemplate.getName())) {
-                    continue;
-                }
-                try {
-                    client().admin().indices().prepareDeleteTemplate(indexTemplate.getName()).get();
-                } catch (IndexTemplateMissingException e) {
-                    // ignore
+    private void wipeAllTemplates(Set<String> exclude, ActionListener<Void> listener) {
+        client().admin().indices().prepareGetTemplates().execute(listener.delegateFailure((f, response) -> {
+            try (RefCountingListener subscribableListener = new RefCountingListener(f)) {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:303]: Do we need a separate inner `RefCountingListener` here or could we re-use the outer one?

**Review by DaveCTurner — COMMENTED** [review]
Seems like a good change but there's quite a mix of styles here, I'd rather we used `SubscribableListener#andThen` and so on to clarify the things that run in sequence still.

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java (hunk 3 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:241]: Good question in 2024 :) I don't know. Maybe let's add a timeout for this stuff in a follow-up? I'll make sure to set the interrupt again here though. Sec :)

Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/TestCluster.java:241]: Actually scratch that ... more than 10s for a test cleanup would be stupid ... lets just use `safeGet` and fix things if this ever fails.

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java (hunk 5 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/TestCluster.java]: Right, reused the outer one now :)

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java (hunk 4 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/TestCluster.java]: Jup, as with the other one, just went for `safeGet` now, 10s have to suffice IMO.

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java (hunk 2 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/TestCluster.java]: I think you meant the other template delete method here? Tried my best to cleanly use the subscribale listener there :) Would be nice to have an equivalent of e.g. `java.util.concurrent.CompletionStage#thenCombine` I think, but I think it's reasonably readable now :) let me know what you think.

File: test/framework/src/main/java/org/elasticsearch/test/TestCluster.java (hunk 1 above)
Comment (original-brownbear) [reply test/framework/src/main/java/org/elasticsearch/test/TestCluster.java]: ++ done

**Comment by original-brownbear** [comment]
FYI: Had to fix one more spot in https://github.com/elastic/elasticsearch/pull/111744/files/1456eac1b6d9aec336e8927cae2bf7f8ccc078e6..f10e665ecb1ed9fb760025b65aa617d3c79a7ec3

**Review by DaveCTurner — COMMENTED** [review]
Still seems a little tangled to me, what do you think about something more like https://github.com/elastic/elasticsearch/compare/main...DaveCTurner:elasticsearch:2024/08/09/111744-possible-simplification

**Comment by original-brownbear** [comment]
@DaveCTurner looks good :) Didn't know about `newForked` shall I merge those changes into here?

**Comment by DaveCTurner** [comment]
Sure, feel free

**Review by DaveCTurner — APPROVED** [review]
image

**Comment by original-brownbear** [comment]
:D thanks David!

**Pull request merged**